### ISSUE TITLE
After upgrading to 3.3 my Instance suddenly has all of my repositories enabled

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#3450]
Before 3.3 my instance worked fine, I had two repositories enabled as shown by this search query:

![image](https://user-images.githubusercontent.com/3173176/56253856-1452ff80-6073-11e9-88ee-f1b233f3b8aa.png)

After upgrading, the same search greets me with:

![image](https://user-images.githubusercontent.com/3173176/56253880-2d5bb080-6073-11e9-8b1f-f90fe9d9b91f.png)

After upgrading my external service configuration is:

```
{
  // Use Ctrl+Space for completion, and hover over JSON properties for documentation.
  // Configuration options are documented here:
  // https://docs.sourcegraph.com/admin/external_service/github#configuration

  "url": "https://github.com",

  // A token is required for access to private repos, but is also helpful for public repos
  // because it grants a higher hourly rate limit to Sourcegraph.
  // Create one with the repo scope at https://[your-github-instance]/settings/tokens/new
  "token": "REDACTED",

  // Sync public repositories from https://github.com by adding them to "repos".
  // (This is not necessary for GitHub Enterprise instances)
  "repos": [
       "sourcegraph/sourcegraph"
  ],
  "repositoryQuery": [
    "affiliated"
  ]
}
```

## Reproducing

1. Start Sourcegraph 3.2.2, add GitHub external service + github token + the following config:

```
"repos": [
       "sourcegraph/sourcegraph"
  ],
```

2. Go to repositories page and enable `sourcegraph/sourcegraph` and `sourcegraph/infrastructure` repositories.
3. Upgrade to Sourcegraph 3.3.0-rc.1, see behavior described above.

**Comment by slimsag** [comment]
I am tagging with release-blocker until we confirm this is behavior we want.

**Comment by sqs** [comment]
This is indeed not the intended behavior:

From https://github.com/sourcegraph/sourcegraph/issues/2025:

> Migrate all disabled repos to github.exclude list.

All of your disabled repositories should have been added to the `exclude` property in your GitHub external service config.

**Comment by keegancsmith** [comment]
This is not the behaviour we want. This is likely a bad bug in our migration logic.

**Comment by keegancsmith** [comment]
Found the bug. This was due to a regression we introduced yesterday afternoon! Who knew adding observability can lead to a bug like this, and given how we added observability it wasn't caught by our unit tests (not all our unit tests have the observability wrappers).

The root cause is some functionality we had which did type assertions on sources (such as github). The problem is we would wrap the sources with an "ObservableSource", causing the type assertions to fail => the migration thought we had no sources to migrate.

I am fixing this now. Out of interest here is my notes / test rig for future generations:

<details>
Running 3.2 against my local DB

```shellsession
docker run --publish 7080:7080 --publish 2633:2633 --rm \
  --volume ~/.sourcegraph/config:/etc/sourcegraph \
  --volume ~/.sourcegraph/data:/var/opt/sourcegraph \
  -e 'PGDATASOURCE=postgres://keegan@docker.for.mac.localhost:5432/keegan?sslmode=disable' \
  sourcegraph/server:3.2.2

pg_dump keegan > keegan-3_2_2-dump.sql

dropdb keegan; createdb keegan; psql keegan < keegan-3_2_2-dump.sql

GITHUB_BASE_URL=https://api.github.com REDIS_ENDPOINT=127.0.0.1:6379 go run main.go
```

Easily reproduced. Gonna add some extra logging to see why.

Not reproducing with my isolated repoupdater

I find it kinda weird that we set both ID and name for the exclude list.

The problem is due to instrumentation wrapping types and us using type
assertions.

```diff
diff --git a/cmd/repo-updater/repos/migrations.go b/cmd/repo-updater/repos/migrations.go
index 4fe99c82e3..55e5ec1bc4 100644
--- a/cmd/repo-updater/repos/migrations.go
+++ b/cmd/repo-updater/repos/migrations.go
@@ -1,9 +1,13 @@
 package repos
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"fmt"
+	"io/ioutil"
 	"net/url"
+	"sort"
 	"time"
 
 	"github.com/pkg/errors"
@@ -56,6 +60,7 @@ func EnabledStateDeprecationMigration(sourcer Sourcer, clock func() time.Time, k
 			sourced, err = srcs.ListRepos(ctx)
 			cancel()
 		}
+		writeRepos("sourced", sourced)
 
 		if err != nil {
 			return errors.Wrapf(err, "%s.sources.list-repos", prefix)
@@ -65,6 +70,7 @@ func EnabledStateDeprecationMigration(sourcer Sourcer, clock func() time.Time, k
 			Kinds:   kinds,
 			Deleted: true,
 		})
+		writeRepos("stored", stored)
 
 		if err != nil {
 			return errors.Wrapf(err, "%s.store.list-repos", prefix)
@@ -80,7 +86,8 @@ func EnabledStateDeprecationMigration(sourcer Sourcer, clock func() time.Time, k
 		svcs := make(map[int64]*service, len(all))
 		upserts := make(ExternalServices, 0, len(all))
 
-		for _, e := range all {
+		for i, e := range all {
+			mustWriteJSON(fmt.Sprintf("all-service-%d", i), e)
 			// Skip any injected sources that are not persisted.
 			if e.ID != 0 {
 				svcs[e.ID] = &service{svc: e}
@@ -140,7 +147,9 @@ func EnabledStateDeprecationMigration(sourcer Sourcer, clock func() time.Time, k
 		}
 
 		now := clock()
-		for _, e := range svcs {
+		for i, e := range svcs {
+			mustWriteJSON(fmt.Sprintf("service-%d", i), e)
+
 			if err = removeInitalRepositoryEnablement(e.svc, now); err != nil {
 				return errors.Wrapf(err, "%s.remove-initial-repository-enablement", prefix)
 			}
@@ -384,3 +393,49 @@ func transact(ctx context.Context, tr Transactor, m Migration) (err error) {
 
 	return m(ctx, tx)
 }
+
+var debugSuffix = time.Now().Format("150405")
+
+func writeRepos(name string, rs Repos) {
+	type debugRepo struct {
+		Name      string    `json:"name,omitempty"`
+		Enabled   bool      `json:"enabled,omitempty"`
+		DeletedAt time.Time `json:"deleted_at,omitempty"`
+	}
+	dr := make([]debugRepo, len(rs))
+	for i, r := range rs {
+		dr[i] = debugRepo{
+			Name:      r.Name,
+			Enabled:   r.Enabled,
+			DeletedAt: r.DeletedAt,
+		}
+	}
+
+	var b bytes.Buffer
+	enc := json.NewEncoder(&b)
+	sort.Slice(dr, func(i, j int) bool {
+		return dr[i].Name < dr[j].Name
+	})
+	for _, r := range dr {
+		err := enc.Encode(&r)
+		if err != nil {
+			panic(err)
+		}
+	}
+	mustWrite(name, b.Bytes())
+}
+
+func mustWriteJSON(name string, v interface{}) {
+	b, err := json.MarshalIndent(v, "", "  ")
+	if err != nil {
+		panic(err)
+	}
+	mustWrite(name, b)
+}
+
+func mustWrite(name string, b []byte) {
+	err := ioutil.WriteFile(fmt.Sprintf("%s-%s.json", debugSuffix, name), b, 0600)
+	if err != nil {
+		panic(err)
+	}
+}
diff --git a/main.go b/main.go
new file mode 100644
index 0000000000..f4fa9da99e
--- /dev/null
+++ b/main.go
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/cmd/repo-updater/repos"
+)
+
+const port = "3182"
+
+func main() {
+	ctx := context.Background()
+
+	clock := func() time.Time { return time.Now().UTC() }
+	dsn := "postgres://keegan@localhost:5432/keegan?sslmode=disable"
+	db, err := repos.NewDB(dsn)
+	if err != nil {
+		log.Fatalf("failed to initialize db store: %v", err)
+	}
+
+	// repo-updater doesn't run DB migrations, but since this runs in
+	// isolation we need it to.
+	if err := repos.MigrateDB(db); err != nil {
+		log.Fatal(err)
+	}
+
+	store := repos.NewDBStore(ctx, db, sql.TxOptions{Isolation: sql.LevelSerializable})
+	var src repos.Sourcer
+	{
+		cf := repos.NewHTTPClientFactory()
+		src = repos.NewSourcer(cf)
+	}
+
+	kinds := []string{
+		"GITHUB",
+		"GITLAB",
+		"BITBUCKETSERVER",
+		"OTHER",
+	}
+	migrations := []repos.Migration{
+		repos.GithubSetDefaultRepositoryQueryMigration(clock),
+		repos.GitLabSetDefaultProjectQueryMigration(clock),
+		repos.EnabledStateDeprecationMigration(src, clock, kinds...),
+	}
+
+	for _, m := range migrations {
+		if err := m.Run(ctx, store); err != nil {
+			log.Fatalf("failed to run migration: %s", err)
+		}
+	}
+
+	svcs, err := store.ListExternalServices(ctx, repos.StoreListExternalServicesArgs{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, s := range svcs {
+		fmt.Println(s.DisplayName, s.Config)
+	}
+
+	repos, err := store.ListRepos(ctx, repos.StoreListReposArgs{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println()
+	fmt.Println(len(repos))
+	for _, r := range repos[:min(len(repos), 5)] {
+		fmt.Println(r.Name)
+	}
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
```
</details>

**Comment by keegancsmith** [comment]
Fix cherry-picked onto 3.3 as 49f0b110029fbdf172e0e57b445bd56ff748b5ce
